File: async/javastore/app/models/Backends.java
package models;

import play.libs.Akka;
import play.libs.F;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

import static play.libs.F.Promise;

public class Backends {

    public Promise<User> getUserById(final String userId) {
        return Akka.future(new Callable<User>() {
            @Override
            public User call() throws Exception {
                return new User(userId, "email@domain.com");
            }
        });
    }

    public Promise<List<Order>> getOrdersForUser(final String email) {
        return Akka.future(new Callable<List<Order>>() {
            @Override
            public List<Order> call() throws Exception {
                List<Order> result = new ArrayList<>();
                result.add(new Order("productA", email));
                return result;
            }
        });
    }

    public Promise<List<Product>> getProductsForOrders(final List<Order> orders) {
        return Akka.future(new Callable<List<Product>>() {
            @Override
            public List<Product> call() throws Exception {
                List<Product> products = new ArrayList<>();
                for (Order order : orders) {
                    products.add(new Product(order.getProductId(), "description"));
                }
                return products;
            }
        });
    }

    public Promise<List<Stock>> getStocksForProducts(final List<Product> products) {
        return Akka.future(new Callable<List<Stock>>() {
            @Override
            public List<Stock> call() throws Exception {
                List<Stock> stocks = new ArrayList<>();
                for (Product product : products) {
                    stocks.add(new Stock(product.getProductId(), 5));
                }
                return stocks;
            }
        });
    }

}

File: async/scalastore/app/models/Backends.scala
package models

import scala.concurrent.Future

case class ProductId(id: String) extends AnyVal
case class Email(value: String) extends AnyVal {
  override def toString: String = value
}

case class User(id: String, email: Email)
case class Order(productId: ProductId, email: Email)
case class Product(productId: ProductId, description: String)
case class Stock(productId: ProductId, quantity: Int)

trait Backends {

  def getUserById(userId: String): Future[User] =
    Future.successful(User(userId, Email("email@domain.com")))

  def getOrdersForUser(email: Email): Future[Seq[Order]] =
    Future.successful(List(Order(ProductId("abc"), email)))

  def getProductsForOrders(orders: Seq[Order]): Future[Seq[Product]] =
    Future.successful(
      for (order <- orders) yield Product(order.productId, "description")
    )

  def getStocksForProducts(products: Seq[Product]): Future[Seq[Stock]] =
    Future.successful(
      for (product <- products) yield Stock(product.productId, 3)
    )

}
